lib/private/GlobalScale/Config.php

```diff
@@ -66,7 +82,127 @@ public function onlyInternalFederation() {
 
 		$enabled = $this->config->getSystemValue('gs.federation', 'internal');
 
-		return $enabled === 'internal';
+		$type = strtolower($type);
+		$typeEnabled = 'internal';
+		if (in_array($type, [IGlobalScaleConfig::INCOMING, IGlobalScaleConfig::OUTGOING])) {
+			$typeEnabled = $this->config->getSystemValue('gs.federation.' . $type, 'internal');
+		}
+
+		return $enabled === 'internal' && $typeEnabled === 'internal';
+	}
+
+
+	/**
+	 * @param string $remote
+	 * @param string $token
+	 * @param string $key
+	 *
+	 * @return bool
+	 * @since 19.0.0
+	 */
+	public function allowedOutgoingFederation(string $remote, string $token = '', string $key = ''): bool {
+		if (!$this->onlyInternalFederation(self::OUTGOING)) {
+			return true;
+		}
+
+		if ($key !== '' && $token !== '') {
+			return $this->keyIsInternal($token, $key);
+		}
+
+		return $this->remoteIsInternal($remote);
+	}
+
+
+	/**
+	 * @param string $remote
+	 * @param string $token
+	 * @param string $key
+	 *
+	 * @return bool
+	 * @since 19.0.0
+	 */
+	public function allowedIncomingFederation(string $remote, string $token, string $key): bool {
+		if (!$this->onlyInternalFederation(self::INCOMING)) {
+			return true;
+		}
+
+		if (!$this->remoteIsInternal($remote)) {
+			return false;
+		}
+
+		return $this->keyIsInternal($token, $key);
+	}
+
+
+	/**
+	 * @param string $token
+	 *
+	 * @return string
+	 */
+	public function generateInternalKey(string $token): string {
+		$jwt = $this->config->getSystemValue('gss.jwt.key', '');
+		if ($jwt === '' || $token === '') {
+			return '';
+		}
+
+		return md5($token . '-' . $jwt);
+	}
+
+
+	/**
+	 * @param string $remote
+	 *
+	 * @return bool
+	 */
+	public function remoteIsInternal(string $remote): bool {
+		if (!$this->isGlobalScaleEnabled()) {
+			return false;
+		}
+
+		$tmp = parse_url($remote, PHP_URL_HOST);
+		$remote = ($tmp === null) ? $remote : $tmp;
+
+		if (in_array($remote, $this->getGSInstances())) {
+			return true;
+		}
+
+		return false;
+	}
+
+
+	/**
+	 * @param string $token
+	 * @param string $key
+	 *
+	 * @return bool
+	 */
+	private function keyIsInternal(string $token, string $key): bool {
+		if ($key === $this->generateInternalKey($token)) {
+			return true;
+		}
+
+		return false;
+	}
+
+
+	/**
+	 * @return array
+	 */
+	public function getGSInstances(): array {
+		/** @var string $lookup */
+		$lookup = $this->config->getSystemValue('lookup_server', '');
+
+		$request = new Request('/instances', Request::TYPE_GET);
+		$request->setAddressFromUrl($lookup);
+
+		try {
+			return $instances = $this->retrieveJson($request);
+		} catch (RequestContentException | RequestNetworkException | RequestResultSizeException | RequestServerException | RequestResultNotJsonException $e) {
+			\OC::$server->getLogger()
+						->log(2, 'Issue while retrieving instances from lookup: ' . $e->getMessage());
```

Comment:
```suggestion
						->warning('Issue while retrieving instances from lookup: ' . $e->getMessage());
```

Or if you want to stick with `log`: 
https://github.com/nextcloud/server/blob/5bf3d1bb384da56adbf205752be8f840aac3b0c5/lib/public/ILogger.php#L54